### PR TITLE
Add support for IPV6_MULTICAST_IF and IP_MULTICAST_IF socket options

### PR DESCRIPTION
**Pull request by jukkar**
Allow user to set the network interface for IPv6 and IPv4 multicast sockets.

### CODE REVIEW

**Review by rlubos — DISMISSED** [review]

File: subsys/net/ip/net_context.c [hunk 1]
```diff
@@ -766,6 +766,18 @@ int net_context_bind(struct net_context *context, const struct sockaddr *addr,
 		if (net_ipv6_is_addr_mcast(&addr6->sin6_addr)) {
 			struct net_if_mcast_addr *maddr;
 
+			if (IS_ENABLED(CONFIG_NET_UDP) &&
+			    (net_context_get_type(context) == SOCK_DGRAM ||
+			     net_context_get_type(context) == SOCK_RAW)) {
```
[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/ip/net_context.c:771]: The Linux manual page talks about `SOCK_RAW` socket but I am not sure what should we do in that case in our code. I will remove the `SOCK_RAW` until that is figured out.

File: tests/net/socket/udp/src/main.c [hunk 2]
```diff
@@ -2531,6 +2534,141 @@ ZTEST(net_socket_udp, test_37_ipv6_src_addr_select)
 				       &my_addr3, &dest);
 }
 
+extern bool loopback_dont_swap_addresses;
```
[the remark this answers is not in the document]
Comment (jukkar) [reply tests/net/socket/udp/src/main.c:2537]: Ok, we can certainly do that. I do not want to expose this publicly, no one except this one test needs this functionality anyway.

**Comment by jukkar** [comment]
* updated according to comments
* rebased against latest main
* added support and tests for `IP_MULTICAST_IF` socket option

**Review by rlubos — DISMISSED** [review]
Looks good, just one suggestion.

**Comment by jukkar** [comment]
* resolved merge conflicts
* rebased to latest main

**Review by rlubos — APPROVED** [review]

**Comment by jukkar** [comment]
@pdgendt if you have time, please take a look, and thanks in advance!

**Review by cfriedt — APPROVED** [review]

**Pull request merged**